pkg/rebalancer/storage/controller/service/rebalancer.go

```diff
@@ -19,28 +19,526 @@ package service
 
 import (
 	"context"
+	"math"
 	"reflect"
+	"strconv"
+	"sync/atomic"
+	"time"
 
+	"github.com/vdaas/vald/internal/errgroup"
 	"github.com/vdaas/vald/internal/errors"
+	"github.com/vdaas/vald/internal/k8s"
+	"github.com/vdaas/vald/internal/k8s/configmap"
+	"github.com/vdaas/vald/internal/k8s/decoder"
+	"github.com/vdaas/vald/internal/k8s/job"
+	mpod "github.com/vdaas/vald/internal/k8s/metrics/pod"
+	"github.com/vdaas/vald/internal/k8s/pod"
+	"github.com/vdaas/vald/internal/k8s/statefulset"
+	"github.com/vdaas/vald/internal/log"
+	"github.com/vdaas/vald/internal/safety"
+	"github.com/vdaas/vald/pkg/rebalancer/storage/controller/config"
+	"github.com/vdaas/vald/pkg/rebalancer/storage/controller/model"
 )
 
+const (
+	qualifiedNamePrefix string = "rebalancer.vald.vdaas.org/"
+)
+
+// Rebalancer represents the rebalancer interface.
 type Rebalancer interface {
-	Start(context.Context) (<-chan error, error)
+	Start(ctx context.Context) (<-chan error, error)
 }
 
-type rebalancer struct{}
+type rebalancer struct {
+	podName      string
+	podNamespace string
+
+	jobs           atomic.Value
+	jobName        string
+	jobNamespace   string
+	jobTemplate    atomic.Value
+	jobTemplateKey string // config map key of job template
+
+	agentName         string
+	agentNamespace    string
+	agentResourceType config.AgentResourceType
+	pods              atomic.Value
+	podMetrics        atomic.Value
+
+	jobConfigmapName string
+
+	statefulSets atomic.Value
+
+	leaderElectionID string
 
-func New(opts ...Option) (dsc Rebalancer, err error) {
+	rcd           time.Duration // reconcile check duration
+	eg            errgroup.Group
+	ctrl          k8s.Controller
+	tolerance     float64
+	rateThreshold float64
+	decoder       *decoder.Decoder
+}
+
+// NewRebalancer initialize job, configmap, pod, podMetrics, statefulset, replicaset, daemonset reconciler.
+// And it returns the rebalancer implemenation or any error occurred.
+func NewRebalancer(opts ...RebalancerOption) (Rebalancer, error) {
```

Comment:
![badge](https://img.shields.io/badge/review-imo-orange.svg?style=flat-square)
Note: there're too many inline functions under this function. It is better to define them as top-level private functions for testability.

Reply:
thank you.
got it. 
I fixed: https://github.com/vdaas/vald/pull/1187/commits/6f57d61963647d4709b267b3d49964aaf67f2f43

Reply:
Not fixed yet.
I mean the functions like 
```
statefulset.WithOnReconcileFunc(func(statefulSetList map[string]) {
...
})
```
should be defined as top-level function for testability:
```
statefulset.WithOnReconcileFunc(fn0)
...
func fn0(statefulSetList map[string]) {
...
}
```

Reply:
@rinx 
I'm sorry, I misunderstood
thank you.
https://github.com/vdaas/vald/pull/1187/commits/51d34852992605ac47845619e024b2840db5be67